src/main/scala/edu/ie3/osmogrid/lv/region_coordinator/OsmoGridModelPartitioner.scala

```diff
@@ -0,0 +1,180 @@
+      enhancedEntities: ParSeq[EnhancedOsmEntity],
+      areas: ParMap[AreaKey, Polygon],
+      allocationStrategy: EntityAllocationStrategy
+  ): ParMap[AreaKey, ParSeq[EnhancedOsmEntity]] = {
+
+    enhancedEntities
+      .flatMap { entity =>
+        assign(entity, areas, allocationStrategy).map(_ -> entity)
+      }
+      .groupBy(_._1)
+      .map { case (areaId, entities) =>
+        areaId -> entities.map(_._2)
+      }
+  }
+
+  private def assign(
+      enhancedEntity: EnhancedOsmEntity,
+      areas: ParMap[AreaKey, Polygon],
+      allocationStrategy: EntityAllocationStrategy
+  ): Iterable[AreaKey] = {
+
+    val entityVotes = vote(enhancedEntity.entity, enhancedEntity, areas)
+
+    allocationStrategy match {
+      case EntityAllocationStrategy.AssignToAll =>
+        entityVotes.keys
+      case EntityAllocationStrategy.AssignByMax =>
+        entityVotes
+          .maxByOption(_._2)
+          .map(_._1)
+          .map(Iterable.single)
+          .getOrElse(Iterable.empty)
+    }
+  }
+
+  private def vote(
+      entity: OsmEntity,
+      enhancedEntity: EnhancedOsmEntity,
+      areas: ParMap[AreaKey, Polygon]
+  ): Map[AreaKey, Int] =
+    entity match {
+      case node: OsmEntity.Node =>
+        vote(node, areas)
+      case way: OsmEntity.Way =>
+        vote(way, enhancedEntity, areas)
+      case relation: OsmEntity.Relation =>
+        vote(relation, enhancedEntity, areas)
+    }
+
+  private def vote(
+      relation: OsmEntity.Relation,
+      enhancedEntity: EnhancedOsmEntity,
+      areas: ParMap[AreaKey, Polygon]
+  ): Map[AreaKey, Int] =
+    relation.members
+      .flatMap { member =>
+        member.relationType match {
+          case RelationMemberType.Node =>
+            enhancedEntity.node(member.id)
+          case RelationMemberType.Way =>
+            enhancedEntity.way(member.id)
+          case RelationMemberType.Relation =>
+            enhancedEntity.relation(member.id)
+          case RelationMemberType.Unrecognized =>
+            None
+        }
+      }
+      .flatMap { entity =>
+        vote(entity, enhancedEntity, areas).toSeq
+      }
+      .groupBy(_._1)
+      .map { case areaId -> votes =>
+        areaId -> votes.map(_._2).sum
+      }
+
+  private def vote(
+      way: OsmEntity.Way,
+      enhancedEntity: EnhancedOsmEntity,
+      areas: ParMap[AreaKey, Polygon]
+  ): Map[AreaKey, Int] =
+    way.nodes
+      .flatMap(enhancedEntity.node)
+      .flatMap { node =>
+        vote(node, areas)
+      }
+      .groupBy(_._1)
+      .map { case areaId -> votes =>
+        areaId -> votes.map(_._2).sum
+      }
+
+  private def vote(
+      node: OsmEntity.Node,
+      areas: ParMap[AreaKey, Polygon]
+  ): Map[AreaKey, Int] = {
+    val point = GeoUtils.buildPoint(node.latitude, node.longitude)
+    areas.iterator
+      .collectFirst {
+        case areaId -> polygon if polygon.covers(point) =>
+          Map(areaId -> 1)
+      }
+      .getOrElse(Map.empty)
+  }
+}
```

Comment:
I like the implementation idea. I think it might be useful though to give a longer explanation above the partition function of how the assignment and "voting mechanism" works as this is at least not obvious without taking a deeper look at the code. 

Also one might consider if it would be faster / more concise if one would build polygons out of the areas and calculate their overlap via the locationtech package and use this for the assignment. Especially if we end up building the polygons later on anyways and are using them (this would make the way polygon cache johannes built very useful)

Reply:
To me the _creation_ of polygon objects is not the problem here, but rather the determination of whether the entity polygon belongs. Currently, we check if all points individually sit inside the boundary. I guess the alternative here would be calculating intersections with all boundaries and comparing their sizes, which sounds computationally more complex - I could be wrong here though. It's definitely something one could investigate.